Best response: calcul, agrégation et vignette

AlexisCochard opened the pull request:
Pour tester : 
```r
devtools::load_all(export_all = FALSE)
rmarkdown::render("vignettes/BOR.Rmd")
browseURL("vignettes/BOR.html")
```

AlexisCochard committed:
Transfert de l'épisode 1

AlexisCochard committed:
suppression des "rate" pour ORR et CBR

AlexisCochard committed:
ajout clopper_pearson dans utils

AlexisCochard committed:
paramétrisation de la longueur des cycles ( les 28 jours)

AlexisCochard committed:
modif suite des "rate"

AlexisCochard committed:
ajout .remove_post_pd()

AlexisCochard committed:
suppression de l'importation de clopper.pearson

AlexisCochard committed:
maj de la doc

AlexisCochard committed:
ajout de class et attr pour rendre le code stylé

AlexisCochard committed:
modif pour avoir la bonne date à la fin

AlexisCochard committed:
ajout pour le cas ou la première mesure est NA

Avant le waterfall plot ne marchait pas en cas de NA donc j'ai modifié pour mettre les NA en 0

DanChaltiel commented on vignettes/ORR.Rmd:
```diff
@@ -0,0 +1,115 @@
+
+# Data Preparation
+
+## Data Overview
```
Pas besoin de subsections, on fait au plus court.

DanChaltiel reviewed (changes requested):
Super boulot, bravo Alexis !
Je te vois progresser entre chaque PR c'est trop cool 😁
Désolé pour le tsunami de commentaires, il y avait plein de choses à dire.
N'hésite pas si je n'ai pas été clair par endroits.

AlexisCochard commented on R/ORR_table.R:
```diff
@@ -0,0 +1,152 @@
+#' Calculate ORR
+#'
+#' @description
+#' The function `ORR_table()` creates a summary table of recist for each possible response.
```
J'ai détaillé un peu plus dans la vignette, est-ce que je le rajoute là aussi du coup ou je mets un message pour dire que la vignette donne plus d'info ?

DanChaltiel commented on R/calc_best_response.R:
```diff
@@ -43,42 +44,123 @@
+                                        labels = c("Complete response", "Partial response",
+                                                   "Stable disease", "Progressive disease", "Not evaluable"))) %>%
+      slice_head(by=subjid) %>%
+      mutate(Overall_Response = ifelse(bestresponse==1 | bestresponse==2, TRUE, FALSE),
+             Clinical_Benefit = ifelse(duree_suivi_max >= 152 | bestresponse==1 | bestresponse==2, TRUE, FALSE)) %>%
```
Je crois que j'ai confondu CBR et DCR, et je me demande si ce n'est pas aussi le cas dans la vignette. A vérifier et voir ce qu'on veut calculer ou pas.

DanChaltiel commented on R/ORR_table.R:
```diff
@@ -0,0 +1,152 @@
+    mutate(Percentage = round(N / total * 100, 1))
+
+  if(show_CBR){
+  summary_df = bind_rows(ORR, response_counts, CBR) %>%
```
À mon sens, il faut imaginer qu'on ne s'arrêtera pas forcément à ces deux-là.
Sinon on peut envisager cette interface par exemple, en remplacement de `show_CBR=TRUE`: 
`ORR_table(include=c("ORR", 'CBR", "DCR", "resp_list"))`
Ça permet à l'utilisateur de choisir ce qu'il veut mettre et dans quel ordre (comme le `variant` des tables d'AE).
Par contre, c'est un peu galère à coder donc on verra plus tard, dans cette PR ou une autre.

AlexisCochard commented on R/ORR_table.R:
```diff
@@ -0,0 +1,152 @@
+      valign(valign = "bottom", part = "header")
+    } else{
+      Best_Response_during_treatment =  Best_Response_during_treatment %>%
+      footnote(i = 1, j = c(4, 1),
```
Pour le DRY, comme en fonction des conditions (confirmed et cbr) pas mal de truc changent quand même j'ai préféré pour le moment bien séparer

AlexisCochard commented on R/calc_best_response.R:
```diff
@@ -43,42 +44,123 @@
-    filter(n_distinct(date)>=2, .by=subjid) %>%
+    .remove_post_pd(resp = response, date = date) %>%
+    mutate(n = n(), .by=subjid) %>%
+    filter(n >= 2) %>%
```
si j'ajoute ça dans `.check_best_resp()` c'est bon ? `df %>%
    filter(n_distinct(date) != n(), .by=subjid) %>%
    grstat_data_warn("Patients with duplicate date of recist evaluations were ignored.",
                     class="check_best_resp_duplic_eval_warning")
  df`

DanChaltiel replied:
Ca devrait marcher, mais j'aurais plutôt écris comme ça : `df %>% filter(n()>1, by=c(subjid, date))`
Pour le message, on n'ignore pas vraiment les patients, il faut juste dire qu'il y a des duplicates.
Aussi, ajouter un check prévient juste l'utilisateur que son input est potentiellement mauvais, mais ça ne change pas qu'il faut mettre un bon `filter` dans `calc_best_reponse`.

AlexisCochard committed:
modif du nom de la fonction orr_table ->aggregate_recist_rates

AlexisCochard committed:
update du namespace

AlexisCochard committed:
maj du calc_best_response

AlexisCochard committed:
ajout d'une deuxième fonction qui utilisera calc_best_response

voir la quelle on garde

AlexisCochard committed:
maj de orr_table suite aux commentaires

AlexisCochard committed:
maj de la vignette orr

AlexisCochard committed:
modif des summarise en count

AlexisCochard committed:
ajout des case_when en fonction interne

AlexisCochard committed:
ajout de  assert_names_exists()

AlexisCochard committed:
petite modif de la vignette pour éviter un message d'erreur

AlexisCochard committed:
assert_names_exists en mieux

AlexisCochard committed:
On garde uniquement la fonction avec calc_best_response séparé

AlexisCochard committed:
maj de la vignette

AlexisCochard committed:
Modif de ORR en BOR pour que ce soit plus explicite

AlexisCochard committed:
maj des noms des fichiers

AlexisCochard committed:
ajout d'un exemple pour modif les flextable

AlexisCochard committed:
Optimisation du flextable pour plus de DRY

AlexisCochard committed:
meilleure utilisation du slice_min

AlexisCochard committed:
pour le moment on follow recist. 1.1 donc modif pour respecter ça

AlexisCochard committed:
delete de trucs pas trop utile

DanChaltiel commented on vignettes/BOR.Rmd:
````diff
@@ -0,0 +1,130 @@
+
+```{r setup, warning=FALSE, message=FALSE}
+## Load libraries
+library(grstat)
````
je mettrais ça dans le 1er chunk ligne 49
Attention, ta vignette ne compilera pas en l'état, il faut ajouter au moins tidyverse (sinon ça plante quand tu utilises `filter()`)
Pas besoin du commentaire non ? On va considérer que les gens savent ce que fait `library()` à ce niveau 😅

DanChaltiel commented on R/calc_best_response.R:
```diff
@@ -43,42 +44,94 @@
       first_date = min_narm(date, na.rm=TRUE),
       min_sum = min_narm(sum, na.rm=TRUE),
       first_sum = sum[date==first_date],
+      response_num = .encode_response(response),
```
Attention, la fonction n'existe plus depuis que tu as mergé avec main.
J'avais unifié les fonctions dans le package dans https://github.com/Oncostat/grstat/commit/470d1b466bd04b251792b57ac6b16bad88b80d4a, maintenant c'est `.recist_to_num()`.
Vérifie un peu (notament en cas de NA) mais normalement ça ne change rien 😊

DanChaltiel replied:
J'avais harmonisé les fonctions RECIST dans https://github.com/Oncostat/grstat/commit/470d1b466bd04b251792b57ac6b16bad88b80d4a, du coup cette fonction n'existe plus dans ta branche depuis que tu as mergé avec `main`.
La fonction est maintenant `.recist_to_num()`.

Si je ne dis pas de bêtise, la seule différence est la gestion du code d'erreur (-99 vs Inf), normalement ça ne devrait rien changer (mais n'hésite pas à vérifier quand même).

DanChaltiel commented on vignettes/BOR.Rmd:
```diff
@@ -0,0 +1,130 @@
+
+As you can see, there are warnings. You can go to `check_recist()` for more information.
+
+# aggregate_recist_rates() function
```
Le lecteur de la vignette cherche une information dans la vignette. Il connait le nom de concepts mais pas du tout le package.
Du coup les titres de sections/sous-sections sont hyper importants (même si `toc=FALSE`).
Tout ça pour dire : il vaut mieux mettre en titre ce que tu vas faire/calculer plutôt que le nom de la fonction.

DanChaltiel commented on vignettes/BOR.Rmd:
```diff
@@ -0,0 +1,130 @@
+
+# Transformation into a flextable
+
+To transfrome our `aggregate_recist_rates` (a data.frame) into a simple flextable, we just have to use the `flextable()` function. But since we wan't a a more specified flextable, we will use the function `as_flextable()` that will take into account the call of our `aggregate_recist_rates()` to create a flextable following a specific model
```
Ah c'est une super bonne idée de parler de `flextable()`, les gens font souvent la confusion!
Par contre je l'aurais tourné autrement : "la bonne façon de faire c'est `as_flextable()`. Attention, si vous utilisez juste `flextable()` vous aurez un truc moche."

DanChaltiel commented on vignettes/BOR.Rmd:
````diff
@@ -0,0 +1,130 @@
+ftbl3 = tbl3 %>% as_flextable()
+ftbl3
+
+```
````
Je pense qu'un seul exemple (le plus complet) suffit, non ?

```r
tbl3 %>%
  as_flextable()%>%
  bold(i=4)%>%
  footnote(i = 3, j = 4,
           value = as_paragraph("test"),
           ref_symbols ="8", part = "body")
```

DanChaltiel reviewed (changes requested):
C'est suuuuuper propre, bravo ! 🤩🤩
Ce n'est vraiment pas une PR facile, il y a le code de la fonction et la vignette en parallèle, courage !

Petits commentaires généraux en plus de ceux dans le code : 

- Il faut que tu testes que ta vignette peut compiler et qu'elle est lisible. J'ai ajouté le code dans le premier post de la PR si besoin.
- Essaie de garder un style de code harmonieux avec le reste du package (je sais que ça peut être galère, désolé) : pas de majuscules dans les noms, retour à la ligne en fin de mutate() si >3 lignes

AlexisCochard committed:
ajout d'un check si date manquante

AlexisCochard committed:
modif du case_when binaire + ajout d'un paramètre use_pharmasug

AlexisCochard committed:
correction nom d'un titre + ajout bonne library

AlexisCochard committed:
update de la vignette

AlexisCochard committed:
modif de la fonction aggregate + flextable pour mettre BOR et CBR à la fin

AlexisCochard committed:
suppression de code inutile

AlexisCochard committed:
mise au bon format du code

AlexisCochard committed:
maj de la doc + suppressions de variables inutiles

DanChaltiel commented on vignettes/BOR.Rmd:
````diff
@@ -0,0 +1,131 @@
+## Unconfirmed response
+
+```{r}
+best_resp = calc_best_response(recist, subjid="subjid", rc_resp="rcresp", rc_date="rcdt", rc_sum = "rctlsum")
````
Va à la ligne, au moins pour `rc_sum`, pour éviter la scrollbar

DanChaltiel commented on R/calc_best_response.R:
```diff
@@ -43,42 +42,109 @@
+                                                   "Stable disease", "Progressive disease", "Not evaluable"))
+      ) %>%
+      mutate(overall_response = response_num==1 | response_num==2,
+             clinical_benefit = duree_suivi_max >= 152 | overall_response) %>%
```
Dans la vignette, tu écris "stable disease lasting at least six months (using a window of +/-1 month for the RECIST date)"
Du coup il ne faudrait pas un `response_num==3` quelque part ?

DanChaltiel commented on R/calc_best_response.R:
```diff
@@ -43,42 +42,109 @@
 calc_best_response = function(data_recist, ...,
                               rc_sum="RCTLSUM", rc_resp="RCRESP", rc_date="RCDT",
                               subjid="SUBJID", exclude_post_pd=TRUE,
```
J'ai pour projet de changer l'interface de toutes les fonctions de `grstat`.
A la place de : 
`rc_sum="RCTLSUM", rc_resp="RCRESP", rc_date="RCDT", subjid="SUBJID"`
Je voudrais :
`cols = c(rc_sum="RCTLSUM", rc_resp="RCRESP", rc_date="RCDT", subjid="SUBJID")`

C'est beaucoup plus facile à maintenir et à utiliser.
On peut définir le vecteur de noms à l'extérieur de l'appel et l'utiliser dans plusieurs fonctions.
Pour la documentation on peut l'écrire avec une liste, comme ça :
```r
#' @param cols a vector with column names inside `calc_best_response()`
#' * `subjid` blabla
#' * `rc_sum` blabla
#' * `blabla` blabla
```

DanChaltiel replied:
exemple d'utilisation finale (que je ferai moi plus tard)
```r
set_grstat_columns = function(...){
  default = c(rc_sum="RCTLSUM", rc_resp="RCRESP", rc_date="RCDT", subjid="SUBJID")
  rtn=default
  #rtn = modifyList(default, ...)
  options(grstat_cols=rtn)
  invisible(NULL)
}
get_grstat_columns = function(){
  getOption("grstat_cols")
}

set_grstat_columns(rc_sum="huhuhi")
get_grstat_columns(rc_sum="huhuhi")


orr = function(x, y, cols=get_grstat_columns()){
  
}
```

DanChaltiel commented on R/calc_best_response.R:
```diff
@@ -43,42 +42,109 @@
-  data_recist %>%
+  na_date = data_recist %>%
+    select(subjid=any_of2(subjid), date=any_of2(rc_date)) %>%
+    distinct() %>%
```
Ce morceau répète la ligne 65-68

DanChaltiel commented on R/calc_best_response.R:
```diff
@@ -43,42 +42,109 @@
+      previous_response_num = lag(response_num),
+      previous_date = lag(date),
+      previous_response_num_2 = lag(response_num,2),
+      delta_date = as.numeric(date - previous_date),
```
Bug rare mais grave : si tu as des dates sous format POSIX (ça arrive) et que la différence est faible, la soustraction donne un résultat aberrant.
```r
a = as.POSIXlt(Sys.time())
b = a-seconds(1)
c = a-days(1)
as.numeric(a-b) #vaut 1
as.numeric(a-c) #vaut 1 aussi
```
Bonne syntaxe = unité explicite : `as.numeric(difftime(a, c, units="days"))` 
En fait la soustraction appelle `difftime(units="auto")`, un peu dangereux...

DanChaltiel reviewed (commented):
Bon, on arrive au bout :-)
Il faut qu'on discute tranquillement des questions d'interface, j'ai une intuition mais je n'ai pas trop de pratique réelle.
Pour la vignette, c'est beaucoup plus clair !
Il faudra juste passer les paragraphes dans une IA pour corriger les typos et la tournure de certaines phrases.
Et il faut aussi qu'on trouve un moyen de faire des tests propres.

AlexisCochard commented on vignettes/BOR.Rmd:
```diff
@@ -0,0 +1,131 @@
+`aggregate_recist_rates()` consist of 2 paramater :
+
+-   `data`, the data.frame containing the best response obtain after using the function `calc_best_response()` (confirmed or unconfirmed)
+-   `show_cbr`, to show the CBR in the results
```
Je l'ai mis en argument car je rajoute aussi une  footnote qui défini le CBR

AlexisCochard committed:
modif de show_CBR pour derived_endpoints

AlexisCochard committed:
indentation

AlexisCochard committed:
modif de la selection des lignes dans le flextable pour prendre direct la ligne voulue

AlexisCochard committed:
suppression de l'OR et CBR du df de calc_best_resp

AlexisCochard committed:
suppression de doublons

AlexisCochard committed:
Nouvelle fonction pour éviter la double fonction

Pour le moment j'ai gardé les anciennes fonctions et mis en commentaire l'ancien code, à supprimer si nouvelle fonction valider

AlexisCochard committed:
code commenté + nouvelle version pour la fonction

AlexisCochard committed:
Update calc_best_response.R

AlexisCochard committed:
Update calc_best_response.R

AlexisCochard committed:
Modif intro de la vignette

AlexisCochard committed:
Update BOR.Rmd

AlexisCochard committed:
sélection des variables voulues

AlexisCochard committed:
pour que ça reste lisible sur la vignette

AlexisCochard committed:
modif du show_CBR + fusion des 2 chunks

AlexisCochard committed:
information concernant le warning

AlexisCochard committed:
ajout du chunck avec eval = false pour le nom des colonnes

AlexisCochard committed:
update de la doc

AlexisCochard committed:
update pour la validation avec le data set copilot

DanChaltiel reviewed (changes requested):
Merci pour ces nouveaux commits 👍 
Je viens de voir qu'on a un problème de définition entre BOR et ORR 😱 
Il reste à faire : 

- [x] implémenter l'arguments `cols`, cf mon commentaire 
- [ ] implémenter les tests avec testthat, en utilisant le code donné plus haut (on peut voir ça ensemble si tu veux)
- [x] glisser le fichier Rmd dans une IA et utiliser le prompt "corrige les typos et améliore l'anglais". Tu peux évidemment aller plus loin et demander "donne-moi ton avis sur cette vignette" si tu veux.

AlexisCochard committed:
modif de BOR en ORR

AlexisCochard committed:
modif BOR en ORR

AlexisCochard committed:
Amélioration du code aggregate

AlexisCochard committed:
ajout de cols = c()

AlexisCochard committed:
nettoyage du code

AlexisCochard committed:
début amélioration de la vignette

AlexisCochard committed:
update du namespace

AlexisCochard committed:
modif par chat gpt avec le prompt "corrige les typos et améliore l'anglais"

AlexisCochard committed:
remise des fonctions qui avaient disparue de utils

AlexisCochard committed:
modif pour mieux caler sur RECIST 1.1 (en attente retour Yves)

AlexisCochard committed:
Update NAMESPACE

AlexisCochard committed:
Ajout de test pour RECIST

AlexisCochard committed:
bien prendre en compte que si PD dans les 42 premiers jours, on a bien PD et non NE

AlexisCochard committed:
amélioration des tests pour best resp

AlexisCochard committed:
je me suis ajouté dans le package

AlexisCochard committed:
maj des test

DanChaltiel commented on R/calc_best_response.R:
```diff
@@ -26,63 +28,141 @@
+                                                  "Stable disease", "Progressive disease", "Not evaluable"))
+    ) %>%
+    mutate(six_months_confirmation = duree_suivi_max >= 183) %>%
+    select(subjid, best_response=response_final, date, target_sum=sum,
+           target_sum_diff_first=diff_first, target_sum_diff_min=diff_min, six_months_confirmation) %>%
```
TODO: ajouter un `apply_labels()` ici
(et potentiellement sur `aggregate_recist_rates()`)

DanChaltiel commented:
Proposition de test à ajouter : 
```r
tibble(
  SUBJID = 145, 
  RCDT = c(today(), today()+28), 
  RCTLSUM = c(100, NA),
  RCRESP = c(NA, "CR"),
) %>% 
  calc_best_response(rc_date="RCDT", rc_sum="RCTLSUM", rc_resp="RCRESP")
```
C'est un bug dans la version de main, ça vaut le coup de vérifier qu'on n'aura jamais de régression

AlexisCochard committed:
nouveau test

AlexisCochard committed:
ajout apply_labels

AlexisCochard committed:
update doc

AlexisCochard committed:
update finale

AlexisCochard committed:
Squashed commit of the following:

commit 652266a518204855888de8d60d46af04b3ef2929
Author: github-actions[bot] <github-actions[bot]@users.noreply.github.com>
Date:   Thu Apr 9 15:06:37 2026 +0000

    Update dev version (Github Actions)

    Bump to v0.1.0.9029
    [skip ci]

commit dd327d0e23f162ec9dd12da4310aa497896dcba3
Author: AnneLourdessamy <anne.lourdessamy@gmail.com>
Date:   Thu Apr 9 17:05:03 2026 +0200

    95_waterfall_plot_bugs (#100)

    * 95_waterfall_plot_bugs

    gestion des NA dans target_sum_diff_first + création de test-waterfall-plot.R

    * select instead of rename, with all variables

    * fct_reorder2(.na_rm=FALSE)

    solves the bug

    * warn in case of NA

    * remove useless group aesthetic

    feels wrong in plotly

    * more versatile y label (not always target_sum_diff_first)

    * use .get_fill_scale() on the right column (not always named best_response)

    * dont check for missing values on shape

    * doc: improve help

    * test: add snapshots

    * tests: add tests on NA and on renaming

    * fix for chk

    ---------

    Co-authored-by: Dan Chaltiel <15105152+DanChaltiel@users.noreply.github.com>

AlexisCochard committed:
Merge branch 'main' into 58-vignette-calcul-des-orr-le-retour

AlexisCochard committed:
test(waterfall): update snapshot

AlexisCochard committed:
comit final

AlexisCochard committed:
maj suite à update ggplot2
